src/main/java/org/dspace/pack/bagit/BagInfoHelper.java

```diff
@@ -0,0 +1,118 @@
+            return new BagInfoHelper();
+        }
+    });
+
+    private final String TAG_KEY = "replicate-bagit.tag";
+    private final String TAG_SUFFIX = ".txt";
+    private final Map<String, Map<String, String>> tagFiles = new HashMap<>();
+
+    @VisibleForTesting
+    protected BagInfoHelper() {
+        loadFromConfiguration();
+    }
+
+    /**
+     * Get the initialized instance of a BagInfoHelper
+     *
+     * @return the static instance
+     */
+    public static BagInfoHelper getInstance() {
+        return supplier.get();
+    }
+
+    /**
+     * Loads the bag-info.txt and any other fields for tag files found under replicate.bag.tag
+     */
+    private void loadFromConfiguration() {
+        final ConfigurationService configurationService = DSpaceServicesFactory.getInstance().getConfigurationService();
+
+        final List<String> keys = configurationService.getPropertyKeys(TAG_KEY);
+
+        // precomplie patterns for when we split strings
+        final Pattern dotSplit = Pattern.compile("\\.");
+        final Pattern hyphenSplit = Pattern.compile("-");
+
+        for (String key : keys) {
+            final String[] split = dotSplit.split(key);
+            final int keyLength = split.length;
+            if (keyLength != 4) {
+                throw new IllegalArgumentException("Key " + key + " has more values than expected! Please format " +
+                                                   "as replicate-bagit.tag.TAG-FILE.TAG-FIELD");
+            }
+
+            // get the filename to use, and check if it needs to have a suffix attached
+            String file = split[keyLength - 2];
+            if (!file.endsWith(TAG_SUFFIX)) {
+                file = file + TAG_SUFFIX;
+            }
+
+            // normalize the field to be formatted in a BagIt style, e.g. Field-Name
+            final String field = split[keyLength - 1];
+            final StringBuilder bagItNormalized = new StringBuilder();
+            for (String fieldPart : hyphenSplit.split(field)) {
+                bagItNormalized.append(Character.toUpperCase(fieldPart.charAt(0)));
+                bagItNormalized.append(fieldPart.substring(1).toLowerCase());
+                bagItNormalized.append("-");
+            }
+            // remove the trailing hyphen
+            bagItNormalized.deleteCharAt(bagItNormalized.length() - 1);
+
+            Map<String, String> tagFields = tagFiles.get(file);
+            if (tagFields == null) {
+                tagFields = new HashMap<>();
+                tagFiles.put(file, tagFields);
+            }
+
+            tagFields.put(bagItNormalized.toString(), configurationService.getProperty(key));
+        }
+    }
+
+    /**
+     * Get a read-only copy of the properties read in
+     *
+     * @return the tag file properties
+     */
+    public Map<String, Map<String, String>> getTagFiles() {
+        if (tagFiles.isEmpty()) {
+            loadFromConfiguration();
+        }
```

Comment:
While I'm not 100% certain here, it looks to me like we are essentially "caching" the configuration into `tagFiles`.  I'm slightly worried about that, as it seems to me we may be blocking the autoreload capability built into the DSpace ConfigurationService.  In other words, DSpace's ConfigurationService autoreloads every ~5 seconds...this allows system admins to modify `cfg` files and see the changes reload automatically without having to restart/reboot Tomcat (or the servlet container).

The way this `BagInfoHelper` is setup seems like it may be caching the configuration, such that, if someone modifies `replicate-bagit.cfg`, the changes would not take effect until the next time you restart Tomcat.  (I might be wrong here though, so it could be worth testing out the behavior.)

Ideally, to me, this configuration should not be cached in any way, but simply reparsed as needed (the reparsing here shouldn't take long at all). Nearly all DSpace configurations support autoreloading, and this seems like it should as well.

Reply:
Yea I wasn't sure how expensive of an operation reading + parsing the configuration would be, and it sounds like it isn't as much of an issue as I thought it might be. This should be pretty straightforward to update since it will only be removing the cached version.